File: features/auth/fastify/typeorm/files/src/entities/refresh-token.ts
import 'reflect-metadata';
import {
  Column,
  CreateDateColumn,
  Entity,
  Index,
  PrimaryGeneratedColumn,
} from 'typeorm';

@Entity({ name: 'refresh_tokens' })
@Index(['user_id'])
@Index(['session_id'])
export class RefreshToken {
  @PrimaryGeneratedColumn('uuid')
  id!: string;

  @Column({ name: 'user_id', type: 'uuid' })
  user_id!: string;

  @Column({ name: 'session_id', type: 'uuid' })
  session_id!: string;

  @Column({ name: 'token_hash', type: 'varchar', length: 64, unique: true })
  token_hash!: string;

  @Column({ name: 'ip_address', type: 'varchar', length: 64, nullable: true })
  ip_address!: string | null;

  @Column({ name: 'user_agent', type: 'text', nullable: true })
  user_agent!: string | null;

  @Column({ name: 'expires_at', type: 'timestamptz' })
  expires_at!: Date;

  @Column({ name: 'revoked_at', type: 'timestamptz', nullable: true })
  revoked_at!: Date | null;

  @Column({ name: 'rotated_to', type: 'uuid', nullable: true })
  rotated_to!: string | null;

  @Column({ name: 'replay_detected_at', type: 'timestamptz', nullable: true })
  replay_detected_at!: Date | null;

  @CreateDateColumn({ name: 'created_at', type: 'timestamptz' })
  created_at!: Date;
}

File: features/auth/go/gorm/files/internal/auth/models/refresh_token.go
package authmodels

import (
	"time"

	"gorm.io/gorm"

	"projx.local/go/internal/uuid"
)

type RefreshToken struct {
	ID               string     `gorm:"primaryKey;type:uuid" json:"id"`
	UserID           string     `gorm:"type:uuid;index;not null" json:"user_id"`
	SessionID        string     `gorm:"type:uuid;index;not null" json:"session_id"`
	TokenHash        string     `gorm:"uniqueIndex;type:varchar(128);not null" json:"-"`
	IPAddress        string     `gorm:"type:varchar(64)" json:"ip_address"`
	UserAgent        string     `gorm:"type:text" json:"user_agent"`
	ExpiresAt        time.Time  `gorm:"not null" json:"expires_at"`
	RevokedAt        *time.Time `json:"revoked_at,omitempty"`
	RotatedTo        *string    `gorm:"type:uuid" json:"-"`
	ReplayDetectedAt *time.Time `json:"-"`
	CreatedAt        time.Time  `json:"created_at"`
	UpdatedAt        time.Time  `json:"updated_at"`
}

func (RefreshToken) TableName() string { return "refresh_tokens" }

func (r *RefreshToken) BeforeCreate(_ *gorm.DB) error {
	if r.ID == "" {
		r.ID = uuid.V4()
	}
	return nil
}
